sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsShuffleHeartbeatManager.scala

```diff
@@ -79,31 +115,71 @@ class RapidsShuffleHeartbeatManager extends Logging {
    * @return `RapidsExecutorUpdateMsg` with new executors, since the last heartbeat was received.
    */
   def executorHeartbeat(id: BlockManagerId): RapidsExecutorUpdateMsg = synchronized {
-    val lastRegistration = lastRegistrationSeen.get(id)
+    val lastRegistration = executorRegistrations.get(id)
     if (lastRegistration == null) {
       throw new IllegalStateException(s"Heartbeat from unknown executor $id")
```

Comment:
Speaking of gracefully handling executors that heartbeat after becoming stale, this doesn't seem like it's gracefully handling it.  Note that the default timeout is only 10 seconds, and a dropped TCP packet could result in a minute or more before it is recovered.  Also when running on an overloaded machine or heavily GC'ing JVM, I could easily see it taking more than 10 seconds, so I would fully expect this to happen in the real world.

Reply:
Yes, thought about this but I didn't go for the change, thanks for raising. I believe the stale executor could rejoin with a new order. Just need to make sure that the suddenly reappearing executor doesn't cause issues at the executor level trying to connect again.